src/avl.py

```diff
@@ -0,0 +1,468 @@
+
+    @property
+    def max_height(self) -> int:
+        """Returns the height of the maximal depth subtree."""
+
+        return max(self.left_height, self.right_height)
+
+    @property
+    def leaf(self) -> bool:
+        """Returns whether this node is a leaf or not."""
+
+        return self.left is None and self.right is None
+
+    def become_branch(self):
+        """Converts this node from a leaf to a branch."""
+
+        self._left = AVLTreeNode(self.container)
+        self._right = AVLTreeNode(self.container)
+        self._left.parent = self._right.parent = self
+        self.fix_height_metadata()
+
+    @property
+    def balance_factor(self) -> int:
+        """This represents how, and in which direction, this node is imbalanced."""
+
+        return self.left_height - self.right_height
+
+    def update_log(self, operation, force=False):
+        """Updates the log of operation and tree pairs for later visualization.
+
+        The log is only ever updated at the root, to avoid visualizing multiple layers
+        in one stream, which would be very confusing.
+
+        Args:
+          operation: A textual description of the operation taking place, intended
+            for human consumption.
+          force: Defaults to false. If this is true, we will append to the log regardless
+            of whether or not we are the root. This is used during operations where the
+            root has just changed, but the whole tree should still be visualized.
+        """
+
+        # if we aren't the root, don't log
+        if not force and (self.container.root != self):
+            return
+
+        self.container.log.append([operation, self.container.root.to_dict()])
+
+    def insert(self, value):
+        """Inserts the given value into the appropriate spot in the tree.
+
+        Args:
+          value: the value to insert.
+        """
+
+        force = self.container._root == self
+
+        def post_insert():
+            self.update_log("after inserting %i" % value, force=force)
+
+        self.update_log("before inserting %i" % value)
+
+        if self.leaf:
+            self.value = value
+            self.become_branch()
+        elif value < self.value:
+            self.left.insert(value)
+        elif value > self.value:
+            self.right.insert(value)
+        else:
+            # need to use a decorator for logging to avoid problems because of early return
+            post_insert()
+            return # ignore duplicates
+
+        self.fix_height_metadata()
+        self.balance()
+
+        post_insert()
+
+    # this could be __getitem__ (same for set_child and __setitem__)
+    def child(self, side):
+        """Returns the child on side (where side is "right" or "left").
+
+        Args:
+          side: One of "right" or "left". The side of the child we want to fetch."""
+
+        assert side in ("left", "right")
+        return getattr(self, "_" + side)
```

Comment:
You should probably do some microbenchmarks to be sure, but my intuition is
- string comparison in the `assert` is slower than int or similar
- concatenation of strings: likely allocation + copy

`__getitem__` could work. As could using static `_left` and `_right` strings instead of concatenation, together with `__slots__` for the attributes.